airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py

```diff
@@ -472,3 +466,261 @@ class Labels(GoogleAdsStream):
+    Stream is only used internally to implement incremental updates for child streams of IncrementalEventsStream
+    """
+
+    cursor_field = "change_status.last_change_date_time"
+    slice_step = pendulum.duration(microseconds=1)
+    days_of_data_storage = 90
+    cursor_time_format = "YYYY-MM-DD HH:mm:ss.SSSSSS"
+
+    def __init__(self, **kwargs):
+        # date range is not used for these streams, only state is used to sync recent records, otherwise full refresh
+        for key in ["start_date", "conversion_window_days", "end_date"]:
+            kwargs.pop(key, None)
+        super().__init__(start_date=None, conversion_window_days=0, end_date=None, **kwargs)
+
+    @property
+    def query_limit(self) -> Optional[int]:
+        "Queries for ChangeStatus resource have to include limit in it"
+        return 10000
+
+    def read_records(
+        self, sync_mode: SyncMode, cursor_field: List[str] = None, stream_slice: MutableMapping[str, Any] = None, **kwargs
+    ) -> Iterable[Mapping[str, Any]]:
+        """
+        This method is overridden to handle GoogleAdsException with EXPIRED_PAGE_TOKEN error code,
+        and update `start_date` key in the `stream_slice` with the latest read record's cursor value, then retry the sync.
+        """
+        while True:
+            records_count = 0
+            customer_id = stream_slice and stream_slice["customer_id"]
+
+            try:
+                # count records to update slice date range with latest record time when limit is hit
+                records = super().read_records(sync_mode, stream_slice=stream_slice)
+                for records_count, record in enumerate(records, start=1):
+                    self._update_state(customer_id, record)
+                    yield record
+            except GoogleAdsException as exception:
+                # stream_slice will be updated in _handle_expired_page_exception if needed
+                self._handle_expired_page_exception(exception, stream_slice, customer_id)
+            else:
+                # if records limit is hit - update slice with new start_date to continue reading
+                if self.query_limit and records_count == self.query_limit:
+                    # if state was not updated before hitting limit - add step to start_date to avoid infinite loop
+                    current_state = self.get_current_state(customer_id, default=stream_slice["start_date"])
+                    stream_slice["start_date"] = current_state
+
+                    if stream_slice["start_date"] == self.get_current_state(customer_id):
+                        raise AirbyteTracedException(
+                            message=f"More then limit {self.mandatory_limit} records with same cursor field. Incremental sync is not possible for this stream.",
+                            failure_type=FailureType.system_error,
+                        )
+                return
```

Comment:
shouldn't this return be inside an `else`? Otherwise we'd update the stream slice then return from the method anyways. 

Could you test this behavior in a unit test?